File: sdks/typescript/test/person-display-name.test.ts
/**
 * The one human label a Person record can always produce.
 *
 * Every consumer of `users()` was writing this join by hand. None of the three
 * fields it reads can be relied on alone. The same cases run in the Java and
 * Python suites.
 *
 * A function here rather than a method, because `Person` is a plain wire
 * interface in this SDK — the same thing Java puts on the record and Python on
 * the dataclass.
 */

import { test } from "node:test";
import assert from "node:assert/strict";

import { displayName } from "../src/types.js";
import type { Person } from "../src/types.js";

function person(firstName: string | null, lastName: string | null, email: string | null): Person {
  return { firstName, lastName, email } as Person;
}

const CASES: Array<[string | null, string | null, string | null, string | null]> = [
  ["Ada", "Lovelace", "ada@example.com", "Ada Lovelace"],
  ["Ada", null, "ada@example.com", "Ada"],
  [null, "Lovelace", "ada@example.com", "Lovelace"],
  ["  Ada  ", "  Lovelace  ", null, "Ada Lovelace"],
  [null, null, "ada@example.com", "ada@example.com"],
  ["", "", "ada@example.com", "ada@example.com"],
  ["   ", "   ", "  ada@example.com  ", "ada@example.com"],
  [null, null, null, null],
  [null, null, "   ", null],
];

for (const [first, last, email, expected] of CASES) {
  test(`displayName(${first}, ${last}, ${email}) -> ${expected}`, () => {
    assert.equal(displayName(person(first, last, email)), expected);
  });
}

test("a name wins over the email", () => {
  // Not "Name <email>". That is a presentation choice, and a caller who wants
  // it composes one — baking it in would make the common case wrong.
  const label = displayName(person("Ada", "Lovelace", "ada@example.com"));
  assert.equal(label, "Ada Lovelace");
  assert.ok(!label!.includes("ada@example.com"));
});

File: sdks/java/fm-sdk/src/test/java/fm/PersonDisplayNameTest.java
package fm;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

/**
 * The one human label a {@link Person} record can always produce.
 *
 * <p>Every consumer of {@code users()} was writing this join by hand — the
 * reference robot in fm-robot-llm-eval, and anything else that has to put a
 * name next to an {@code ownerId}. None of the three fields it reads can be
 * relied on alone.
 *
 * <p>The same cases run in the Python and TypeScript suites. They are written
 * out rather than shared as a fixture because this is a pure function over one
 * record, not a sequence of updates over an aggregator — the behaviour fixtures
 * drive {@code MarketBook} and {@code Trades} and have nowhere to put it.
 */
class PersonDisplayNameTest {

    private static Person person(String firstName, String lastName, String email) {
        return new Person(null, null, 1L, 1L, firstName, lastName, email, null, null);
    }

    @ParameterizedTest(name = "[{0}] [{1}] [{2}] -> [{3}]")
    @CsvSource(nullValues = "null", value = {
        "Ada,        Lovelace,       ada@example.com,     Ada Lovelace",
        "Ada,        null,           ada@example.com,     Ada",
        "null,       Lovelace,       ada@example.com,     Lovelace",
        "'  Ada  ',  '  Lovelace  ', null,                Ada Lovelace",
        "null,       null,           ada@example.com,     ada@example.com",
        "'',         '',             ada@example.com,     ada@example.com",
        "'   ',      '   ',          '  ada@example.com ',ada@example.com",
        "null,       null,           null,                null",
        "null,       null,           '   ',               null",
    })
    void displayName(String first, String last, String email, String expected) {
        assertThat(person(first, last, email).displayName()).isEqualTo(expected);
    }

    @Test
    void aNameWinsOverTheEmail() {
        // Not "Name <email>". That is a presentation choice, and a caller who
        // wants it composes one — baking it in would make the common case wrong.
        String label = person("Ada", "Lovelace", "ada@example.com").displayName();

        assertThat(label).isEqualTo("Ada Lovelace").doesNotContain("ada@example.com");
    }
}
